contracts/poap/src/msg.rs

```diff
@@ -0,0 +1,398 @@
+                start: self.event_info.start_time.to_owned(),
+                end: self.event_info.end_time.to_owned(),
+            });
+        }
+
+        // Check per address limit
+        if self.event_info.per_address_limit == 0 {
+            return Err(ContractError::InvalidPerAddressLimit {});
+        }
+
+        // Check that the poap uri is a valid IPFS url
+        let poap_uri = Url::parse(&self.event_info.base_poap_uri)
+            .map_err(|_err| ContractError::InvalidPoapUri {})?;
+        if poap_uri.scheme() != "ipfs" {
+            return Err(ContractError::InvalidPoapUri {});
+        }
+
+        // Check that the event uri is a valid IPFS url
+        let event_uri = Url::parse(&self.event_info.event_uri)
+            .map_err(|_err| ContractError::InvalidEventUri {})?;
+        if event_uri.scheme() != "ipfs" {
+            return Err(ContractError::InvalidEventUri {});
+        }
+
+        Ok(())
+    }
+}
+
+impl ExecuteMsg {
+    /// Checks that the data inside the message are coherent.
+    /// NOTE: This function don't checks if the address are valid.
+    pub fn validate(&self) -> Result<(), ContractError> {
+        match &self {
+            ExecuteMsg::UpdateEventInfo {
+                start_time,
+                end_time,
+            } => {
+                if start_time >= end_time {
+                    Err(ContractError::StartTimeAfterEndTime {
+                        start: start_time.to_owned(),
+                        end: end_time.to_owned(),
+                    })
+                } else {
+                    Ok(())
+                }
+            }
+            _ => Ok(()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::msg::{EventInfo, ExecuteMsg, InstantiateMsg};
+    use crate::ContractError;
+    use cosmwasm_std::Timestamp;
+    use cw721_base::InstantiateMsg as Cw721InstantiateMsg;
+
+    #[test]
+    fn instantiate_with_invalid_time_error() {
+        // Test start time after end time
+        let start = Timestamp::from_seconds(2);
+        let end = Timestamp::from_seconds(1);
+        let msg = InstantiateMsg {
+            admin: "".to_string(),
+            minter: "".to_string(),
+            cw721_code_id: 0u64.into(),
+            cw721_initiate_msg: Cw721InstantiateMsg {
+                name: "".to_string(),
+                minter: "".to_string(),
+                symbol: "".to_string(),
+            },
+            event_info: EventInfo {
+                creator: "".to_string(),
+                start_time: start.clone(),
+                end_time: end.clone(),
+                per_address_limit: 1,
+                base_poap_uri: "ipfs://domain.com".to_string(),
+                event_uri: "ipfs://domain.com".to_string(),
+            },
+        };
+
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime { start, end },
+            msg.validate().unwrap_err()
+        );
+
+        // Test start time eq end time
```

Comment:
I think it can be splitted to another test something like `instantiate_with_invalid_time_start_time_eq_end_time_error`.